[inductor] Handle nested tuple/list output in fallback kernel

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88495

Summary: Currently fallback kernel in inductor assumes its output is
either a tensor or a tuple/list of tensors. This PR makes it handle more
generic output data structure.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10

desertfire committed:
[inductor] Handle nested tuple/list output in fallback kernel

Summary: Currently fallback kernel in inductor assumes its output is
either a tensor or a tuple/list of tensors. This PR makes it handl more
generic output data structure.

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Handle nested tuple/list output in fallback kernel"


Summary: Currently fallback kernel in inductor assumes its output is
either a tensor or a tuple/list of tensors. This PR makes it handl more
generic output data structure.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

desertfire commented:
This is to fix an internal case. I found it hard to create a unit test for this in OSS.

desertfire committed:
Update on "[inductor] Handle nested tuple/list output in fallback kernel"


Summary: Currently fallback kernel in inductor assumes its output is
either a tensor or a tuple/list of tensors. This PR makes it handle more
generic output data structure.

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Handle nested tuple/list output in fallback kernel"


Summary: Currently fallback kernel in inductor assumes its output is
either a tensor or a tuple/list of tensors. This PR makes it handle more
generic output data structure.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

eellison commented on torch/_inductor/ir.py:
```diff
@@ -2917,45 +2917,41 @@ def create(cls, kernel, *args, **kwargs):
-            )
-            return [
-                (
+        def generate_output(output, index=""):
```
could add asserts that the length of the output at runtime is the same as we observed here ?

desertfire replied:
We do have `assert_size_stride` generated and that will trigger an issue if the output indexing is OOB (that's how I debugged one test failure), so I guess I will skip adding another assert for now.

eellison replied:
that'll catch when the list runtime list is smaller but it won't catch when the runtime list is larger

eellison reviewed (commented):
looking good, one comment. For test maybe `aten::unique_dim` ?

ngimel commented on torch/_inductor/ir.py:
```diff
@@ -2917,45 +2917,41 @@ def create(cls, kernel, *args, **kwargs):
-            return [
-                (
+        def generate_output(output, index=""):
+            if isinstance(output, list):
```
nit
```
if isinstance(output, (list, tuple)):
    return type(output)(...)
```

ngimel commented on torch/_inductor/ir.py:
```diff
@@ -2917,45 +2917,41 @@ def create(cls, kernel, *args, **kwargs):
+                    for i in range(len(output))
+                )
+            else:
+                return (
```
should you assert on `output` type here, before trying to access its device and dtype?

desertfire committed:
Update on "[inductor] Handle nested tuple/list output in fallback kernel"


Summary: Currently fallback kernel in inductor assumes its output is
either a tensor or a tuple/list of tensors. This PR makes it handle more
generic output data structure.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Handle nested tuple/list output in fallback kernel"


Summary: Currently fallback kernel in inductor assumes its output is
either a tensor or a tuple/list of tensors. This PR makes it handle more
generic output data structure.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Handle nested tuple/list output in fallback kernel"


Summary: Currently fallback kernel in inductor assumes its output is
either a tensor or a tuple/list of tensors. This PR makes it handle more
generic output data structure.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Handle nested tuple/list output in fallback kernel"


Summary: Currently fallback kernel in inductor assumes its output is
either a tensor or a tuple/list of tensors. This PR makes it handle more
generic output data structure.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

desertfire commented:
@pytorchbot merge -g
